pkg/cli/admin/mustgather/mustgather.go

```diff
@@ -412,65 +479,16 @@ func (o *MustGatherOptions) Run() error {
 	defer o.logTimestamp()
 
 	var wg sync.WaitGroup
-	wg.Add(len(pods))
 	errCh := make(chan error, len(pods))
+	podChan := make(chan *corev1.Pod)
+	wg.Add(concurrentMG)
+	for i := 0; i < concurrentMG; i++ {
```

Comment:
I'm not comfortable with the way we handle this concurrency pattern and I'd like to spend more time to investigate this. Meanwhile, could you please explain;

For example, we have 5 images in our slice and according to this loop we have processed the first 4 in a separate goroutine. Where is iterating to process the 5th image?.

Reply:
I think, we can achieve always running the max number of pods at the same time by using the buffered channels. Please let me know what you think https://go.dev/play/p/fv6cGIqxG3W

Reply:
Something like this;
```go
...
concurrentChan := make(chan struct{}, concurrentMG)
for _, pod := range pods {
wg.Add(1)
go func(pod *corev1 Pod({
   defer wg.Done()
   concurrentChan <- struct{}{}
   // runWorker
   <-concurrentChan
})
wg.Wait()
close(errCh)
...
```

Reply:
> I'm not comfortable with the way we handle this concurrency pattern and I'd like to spend more time to investigate this.

Please let me know if I could do anything to help here.



> For example, we have 5 images in our slice and according to this loop we have processed the first 4 in a separate goroutine. Where is iterating to process the 5th image?.

The iteration is happening in:
```go
for _, pod := range pods {
    podChan <- pod
}
```

> I think, we can achieve always running the max number of pods at the same time by using the buffered channels. Please let me know what you think https://go.dev/play/p/fv6cGIqxG3W

Yes, we can. And it's _almost_ the same as what I had earlier except that I had a maximum `concurrentMG` number of goroutines running at any given time, while your suggestion has `len(pods)` number of goroutines spinning up immediately and only `concurrentMG` number of goroutines being processed at a given time. If we go with semaphore pattern, I'd prefer former against latter.

However, @ingvagabund suggested implementing worker pool pattern (which you see in the code right now).

I'd prefer either the semaphore pattern I had earlier or this worker pool pattern, since it spins up maximum `concurrentMG` goroutines at any given time. But I'll let you folks decide and let me know which one you are comfortable with maintaining.

Reply:
> which code portion runs the 5th image;
> 
> ```go
> for _, pod := range pods {
>     podChan <- pod # who is sending the pod for 5th image?.
> }
> ```

OK, let me try. The two code blocks below are adding the number of Pods to be created and processed:
```go
pods = append(pods, o.newPod(node.Name, image, hasMaster))
pods = append(pods, o.newPod(o.NodeName, image, hasMaster))
```

So if there are 5/10/20 pods, all of them are being put on the `podChan` for processing. Some code needs to read from this `podChan` and then do the processing.

> Because iteration in pod creation is limited by concurrentMG;
> 
> ```go
> for i := 0; i < concurrentMG; i++ {
> ```

This code is now only limiting the number of worker goroutines, not the number of pods to spin up. Within this `for` loop, we call `o.runWorker`. And within `o.runWorker`, we read from the above `podChan` till the channel is empty. So the 5th Pod that you're talking about gets processed by `runWorker` method because it reads the channel till there's nothing left in it to read.

Reply:
PTAL at this sample program of smaller size and see if it makes sense - https://go.dev/play/p/_jNPiZilviv

Reply:
@ardaguclu if I read your example in https://go.dev/play/p/fv6cGIqxG3W correctly it runs 7 goroutines but only 4 of them do an actual work of processing pods.

Reply:
Adding `fmt.Printf("Running %v goroutine\n", idx)` line into https://go.dev/play/p/fv6cGIqxG3W example through https://go.dev/play/p/7M9ATLUTkjk:
```
Running 6 goroutine
Running 2 goroutine
Running 0 goroutine
Running 1 goroutine
Running 4 goroutine
Running 3 goroutine
Running 5 goroutine
chan is released 2009-11-10 23:00:01 +0000 UTC m=+1.000000001
Function 1
chan is released 2009-11-10 23:00:01 +0000 UTC m=+1.000000001
Function 7
chan is released 2009-11-10 23:00:01 +0000 UTC m=+1.000000001
Function 3
chan is released 2009-11-10 23:00:01 +0000 UTC m=+1.000000001
Function 2
chan is released 2009-11-10 23:00:02 +0000 UTC m=+2.000000001
Function 5
chan is released 2009-11-10 23:00:02 +0000 UTC m=+2.000000001
Function 4
chan is released 2009-11-10 23:00:02 +0000 UTC m=+2.000000001
Function 6
```

Reply:
https://go.dev/play/p/F0pzLGL878- for showing multiple goroutines looping over the same channel:
```
Spinning up a goroutine...
Spinning up a goroutine...
(1) 0
(0) 1
(0) 2
(1) 3
(0) 4
Done
```

Reply:
Marking this resolved since we're going with different approach now.